src/librustc/metadata/decoder.rs

```diff
@@ -268,10 +268,24 @@ fn item_trait_ref<'tcx>(doc: rbml::Doc, tcx: &ty::ctxt<'tcx>, cdata: Cmd)
     doc_trait_ref(tp, tcx, cdata)
 }
 
-fn enum_variant_ids(item: rbml::Doc, cdata: Cmd) -> Vec<ast::DefId> {
-    reader::tagged_docs(item, tag_items_data_item_variant)
-        .map(|p| translated_def_id(cdata, p))
-        .collect()
+struct EnumVariantIds<'a> {
+    iter: reader::TaggedDocsIterator<'a>,
+    cdata: Cmd<'a>,
+}
+
+impl<'a> Iterator for EnumVariantIds<'a> {
+    type Item = ast::DefId;
+
+    fn next(&mut self) -> Option<ast::DefId> {
+        self.iter.next().map(|p| translated_def_id(self.cdata, p))
+    }
+}
```

Comment:
Is it possible to use a better size_hint? It's important since the iterator is collected when used.